lib/bloc/user_profile/user_profile_bloc.dart

```diff
@@ -123,6 +142,28 @@ class UserProfileBloc {
     return user;
   }
 
+  void _listenUserActions() {
+    _userActionsController.stream.listen((action) {
+      var handler = _actionHandlers[action.runtimeType];
+      if (handler != null) {
+        handler(action).catchError((e) => action.resolveError(e));
+      }
+    });
+  }
+
+  Future _updateSecurityModel(UpdateSecurityModel updateSecurityModelAction) async {
+    if (updateSecurityModelAction.newModel.pinCode == null) {
+      await _secureStorage.delete(key: 'pinCode');
+    } else if (updateSecurityModelAction.newModel.pinCode != _currentUser.securityModel.pinCode) {
+      // Write to storage if the new pin code is different from current pin code
+      await _secureStorage.write(key: 'pinCode', value: updateSecurityModelAction.newModel.pinCode);
+    }
+    SecurityModel securityModel = _currentUser.securityModel.copyWith(
+        pinCode: updateSecurityModelAction.newModel.pinCode, secureBackupWithPin: updateSecurityModelAction.newModel.secureBackupWithPin);
+    _userController.add(_currentUser.copyWith(securityModel: securityModel));
```

Comment:
Why not just push the new model?
```suggestion
    _userController.add(_currentUser.copyWith(securityModel: updateSecurityModelAction.newModel));
```